File: scripts/verify_urls.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Phase 5 verification: compares the Hugo build (public/) against the golden
# Jekyll contract in migration/.
#
# Checks:
#   1. Canonical pages: every golden page URL must exist in the Hugo build.
#   2. Aliases: every golden alias URL must exist as an alias page in Hugo.
#   3. Files: media/asset parity.
#
# Usage: ruby scripts/verify_urls.rb

require "set"

MIGRATION = File.expand_path("../migration", __dir__)
PUBLIC = File.expand_path("../public", __dir__)

golden_pages = File.readlines(File.join(MIGRATION, "golden_pages.txt")).map(&:chomp).to_set
golden_aliases = File.readlines(File.join(MIGRATION, "golden_aliases.txt")).map(&:chomp).map { |l| l.split("\t") }
golden_files = File.readlines(File.join(MIGRATION, "golden_files.txt")).map(&:chomp).to_set

# Enumerate the Hugo build the same way the golden set was extracted from _site.
hugo_pages = Set.new
hugo_files = Set.new
Dir.glob("#{PUBLIC}/**/*").sort.each do |path|
  rel = path.sub("#{PUBLIC}/", "")
  next if File.directory?(path)
  hugo_files << rel
  hugo_pages << "/#{rel}" if rel.end_with?(".html", ".xml", ".json") && !rel.end_with?("index.html")
end
Dir.glob("#{PUBLIC}/**/index.html").sort.each do |path|
  hugo_pages << "/#{path.sub("#{PUBLIC}/", "").sub("index.html", "")}"
end

missing_pages = golden_pages - hugo_pages
extra_pages = hugo_pages - golden_pages

alias_urls = golden_aliases.map(&:first).to_set
missing_aliases = alias_urls - hugo_pages

# Files: golden was the Jekyll _site inventory. Some plugin artifacts are consciously
# not reproduced (documented in the migration report).
skip_files = [/^https:/]                                     # broken jekyll-feed artifact (garbage on live site)

missing_files = golden_files - hugo_files - skip_files.map { |r| golden_files.grep(r) }.flatten.to_set
extra_files = hugo_files - golden_files

puts "== pages =="
puts "golden: #{golden_pages.size}  hugo: #{hugo_pages.size}"
puts "MISSING in Hugo (#{missing_pages.size}):"
missing_pages.sort.first(30).each { |u| puts "  #{u}" }
puts "EXTRA in Hugo (#{extra_pages.size}):"
extra_files_pages = extra_pages.sort.first(30)
extra_files_pages.each { |u| puts "  #{u}" }

puts "== aliases =="
puts "golden: #{alias_urls.size}  missing: #{missing_aliases.size}"
missing_aliases.sort.first(30).each { |u| puts "  MISSING #{u}" }

puts "== files =="
puts "golden: #{golden_files.size}  hugo: #{hugo_files.size}"
puts "MISSING in Hugo (#{missing_files.size}):"
missing_files.sort.first(30).each { |u| puts "  #{u}" }
puts "EXTRA in Hugo (#{extra_files.size}):"
extra_files.sort.first(30).each { |u| puts "  #{u}" }

exit(missing_pages.empty? && missing_aliases.empty? && missing_files.empty? ? 0 : 1)

File: scripts/verify_urls.mjs
#!/usr/bin/env node

// Phase 5 verification: compares the Hugo build (public/) against the golden
// Jekyll contract in migration/.
//
// Checks:
//   1. Canonical pages: every golden page URL must exist in the Hugo build.
//   2. Aliases: every golden alias URL must exist as an alias page in Hugo.
//   3. Files: media/asset parity.
//
// Usage: node scripts/verify_urls.mjs

import { readFileSync, readdirSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const MIGRATION = resolve(HERE, "../migration");
const PUBLIC = resolve(HERE, "../public");

// Sort by UTF-8 bytes so output ordering is stable and locale-independent.
function compareBytes(a, b) {
  return Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8"));
}

function readLines(path) {
  const lines = readFileSync(path, "utf8").split("\n");
  if (lines.length > 0 && lines[lines.length - 1] === "") lines.pop();
  return lines;
}

function walk(dir) {
  const found = [];
  let entries;
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch {
    return found;
  }
  for (const entry of entries) {
    if (entry.name.startsWith(".")) continue;
    const full = join(dir, entry.name);
    if (entry.isDirectory()) found.push(...walk(full));
    else found.push(full);
  }
  return found;
}

const goldenPages = new Set(readLines(join(MIGRATION, "golden_pages.txt")));
const goldenAliases = readLines(join(MIGRATION, "golden_aliases.txt")).map(
  (line) => line.split("\t"),
);
const goldenFiles = new Set(readLines(join(MIGRATION, "golden_files.txt")));

// Enumerate the Hugo build the same way the golden set was extracted from _site.
const hugoPages = new Set();
const hugoFiles = new Set();
for (const path of walk(PUBLIC).sort(compareBytes)) {
  const rel = relative(PUBLIC, path);
  hugoFiles.add(rel);
  if (/\.(html|xml|json)$/.test(rel) && !rel.endsWith("index.html")) {
    hugoPages.add(`/${rel}`);
  }
}
for (const path of walk(PUBLIC).sort(compareBytes)) {
  const rel = relative(PUBLIC, path);
  if (rel === "index.html" || rel.endsWith("/index.html")) {
    hugoPages.add(`/${rel.replace(/index\.html$/, "")}`);
  }
}

const missingPages = [...goldenPages].filter((p) => !hugoPages.has(p));
const extraPages = [...hugoPages].filter((p) => !goldenPages.has(p));

const aliasUrls = new Set(goldenAliases.map(([from]) => from));
const missingAliases = [...aliasUrls].filter((u) => !hugoPages.has(u));

// Files: golden was the Jekyll _site inventory. Some plugin artifacts are
// consciously not reproduced (documented in the migration report).
const skipFiles = [/^https:/]; // broken jekyll-feed artifact (garbage on live site)
const skipped = new Set();
for (const pattern of skipFiles) {
  for (const file of goldenFiles) if (pattern.test(file)) skipped.add(file);
}
const missingFiles = [...goldenFiles].filter(
  (f) => !hugoFiles.has(f) && !skipped.has(f),
);
const extraFiles = [...hugoFiles].filter((f) => !goldenFiles.has(f));

const show = (label, values, render) => {
  console.log(`${label} (${values.length}):`);
  for (const value of values.sort(compareBytes).slice(0, 30)) {
    console.log(`  ${render(value)}`);
  }
};

console.log("== pages ==");
console.log(`golden: ${goldenPages.size}  hugo: ${hugoPages.size}`);
show("MISSING in Hugo", missingPages, (u) => u);
show("EXTRA in Hugo", extraPages, (u) => u);

console.log("== aliases ==");
console.log(`golden: ${aliasUrls.size}  missing: ${missingAliases.length}`);
for (const url of missingAliases.sort(compareBytes).slice(0, 30)) {
  console.log(`  MISSING ${url}`);
}

console.log("== files ==");
console.log(`golden: ${goldenFiles.size}  hugo: ${hugoFiles.size}`);
show("MISSING in Hugo", missingFiles, (f) => f);
show("EXTRA in Hugo", extraFiles, (f) => f);

process.exit(
  missingPages.length === 0 && missingAliases.length === 0 && missingFiles.length === 0
    ? 0
    : 1,
);
